src/main/scala/org/broadinstitute/dsde/workbench/sam/service/UserService.scala

```diff
@@ -26,20 +25,116 @@ class UserService(val directoryDAO: DirectoryDAO, val cloudExtensions: CloudExte
+    }
+  }
+
+  private def validateUser(user: SamUser): Option[Seq[ErrorReport]] =
+    Option(Seq(
+        validateUserIds(user),
+        validateEmail(user.email, blockedEmailDomains)
+      ).flatten // flatten to get rid of Nones
+    ).filter(_.nonEmpty) // If the final Seq is empty, filter it out and just return a None
+
+  // user record has to have a GoogleSubjectId and/or an AzureB2CId
+  private def validateUserIds(user: SamUser): Option[ErrorReport] =
+    if(user.googleSubjectId.isEmpty && user.azureB2CId.isEmpty) {
+      Option(ErrorReport("cannot create user when neither google subject id nor azure b2c id exists"))
+    } else None
+
+  private def validateEmail(email: WorkbenchEmail, blockedEmailDomains: Seq[String]): Option[ErrorReport] =
+    if(!UserService.emailRegex.matches(email.value)) {
+      Option(ErrorReport(StatusCodes.BadRequest, s"invalid email address [${email.value}]"))
+    } else if(blockedEmailDomains.exists(domain => email.value.endsWith("@" + domain) || email.value.endsWith("." + domain))) {
+      Option(ErrorReport(StatusCodes.BadRequest, s"email domain not permitted [${email.value}]"))
+    } else None
+
+  // Try to find user by GoogleSubject, AzureB2CId
+  // A registered user is one that has a record in the database and has a Cloud Identifier specified
+  private def tryToFindUserByCloudId(user: SamUser, samRequestContext: SamRequestContext): IO[Option[SamUser]] =
+    openTelemetry.time("api.v1.user.tryToFindUserByCloudId.time", API_TIMING_DURATION_BUCKET) {
+      if (user.googleSubjectId.nonEmpty) {
+        directoryDAO.loadUserByGoogleSubjectId(user.googleSubjectId.get, samRequestContext)
+      } else if (user.azureB2CId.nonEmpty) {
+        directoryDAO.loadUserByAzureB2CId(user.azureB2CId.get, samRequestContext)
+      } else {
+        IO(None)
+      }
+    }
+
+  private def assertUserIsNotAlreadyRegistered(user: SamUser, samRequestContext: SamRequestContext): IO[SamUser] =
+    tryToFindUserByCloudId(user, samRequestContext).flatMap {
+      case Some(registeredUser) => IO.raiseError(new WorkbenchExceptionWithErrorReport(ErrorReport(StatusCodes.Conflict, s"user ${registeredUser.email} is already registered")))
+      case None => IO(user)
+    }
+
+  // TODO: Add a simple "updateUser" method to directoryDAO so we can do all this in one call and return the updated user record
+  private def updateUser(existingUserId: WorkbenchUserId, user: SamUser, samRequestContext: SamRequestContext): IO[SamUser] =
+    openTelemetry.time("api.v1.user.updateUser.time", API_TIMING_DURATION_BUCKET) {
+      for {
+        _ <- if (user.googleSubjectId.nonEmpty) directoryDAO.setGoogleSubjectId(existingUserId, user.googleSubjectId.get, samRequestContext) else IO.unit
+        _ <- if (user.azureB2CId.nonEmpty) directoryDAO.setUserAzureB2CId(existingUserId, user.azureB2CId.get, samRequestContext) else IO.unit
+      } yield user.copy(id = existingUserId)
+    }
```

Comment:
If we want one of these to succeed regardless of whether the other succeeds we should add a unit test for that. I think if one fails the for yield will short circuit.

Reply:
Also if possible we should override the user id higher up in the call stack.

Reply:
I think it is ok if this short circuits if one of the updates fails.  Per the comment, ideally, I would like to have a vanilla `updateUser` method that just updates both of these fields (and the whole user object) in one sql statement, but we don't have that right now.  If we had that though, it would be a "both succeed or nothing gets updated" scenario.  The catch there is that expanding this refactor into the `DirectoryDAO` I think blows up the scope a bit.  I'm  happy to do it if we feel like it makes sense for this PR.  Or I am happy to merge this as it is and add that new method later.  What do you think?